### PR TITLE
Make use of client library to get the connection string

### PR DESCRIPTION
**Pull request by johscheuer**
# Description

The multi-version client library should be able to automatically detect the correct version to communicate with the FDB cluster.

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

## Discussion

I believe this is in general the right direction to use the client library where possible instead of having our code try multiple different configurations.

## Testing

Local testing and unit tests.

## Documentation

-

## Follow-up

-

### CODE REVIEW

**Review by johscheuer — COMMENTED** [review]
The next step to this would be to remove the `getConnectionString`  request (or at least deprecate it and don't use it) and make use of the connection string in the status. In all cases we fetch the status in addition to the connection string, by combining those requests we can reduce the number of FDB calls.

File: controllers/update_status.go [hunk 1]
```diff
@@ -306,49 +305,44 @@ func optionList(options ...string) []string {
...
 	defer func() { cluster.Status.ConnectionString = originalConnectionString }()
 
-	for _, version := range versions {
-		for _, connectionString := range connectionStrings {
-			logger.Info("Attempting to get connection string from cluster",
-				"version", version, "connectionString", connectionString)
-			cluster.Status.RunningVersion = version
-			cluster.Status.ConnectionString = connectionString
-			adminClient, clientErr := r.getDatabaseClientProvider().GetAdminClient(cluster, r)
+	for _, connectionString := range connectionStrings {
```
Comment (jleach4) [inline controllers/update_status.go:322]: 1 at a time?

**Review by jleach4 — APPROVED** [review]

File: controllers/update_status.go (hunk 1 above)
Comment (johscheuer) [reply controllers/update_status.go:322]: We could run them in a go routine but in the current implementation of the admin client we would work over the list in a serial manner anyway (and I'm not sure how go routines and the FDB bindings will work out together). In general we should have two different connection strings only for a short amount of time.

**Comment by johscheuer** [comment]
The failure is related to the velocity test. I opened an issue to investigate that and I try to fix the test case today.

**Review by sears — DISMISSED** [review]
This makes some parts of #1423 redundant.

**Review by johscheuer — COMMENTED** [review]
Ping @ammolitor, could you check why @sears and @saintstack are not able to "approve" PRs?

File: controllers/update_status.go [hunk 2]
```diff
@@ -306,49 +305,44 @@ func optionList(options ...string) []string {
...
-					"version", version, "connectionString", activeConnectionString)
-				return version, activeConnectionString, nil
-			}
-			logger.Error(err, "Error getting connection string from cluster",
-				"version", version, "connectionString", connectionString)
+		activeConnectionString, err := adminClient.GetConnectionString()
+
+		clientErr = adminClient.Close()
+		if clientErr != nil {
+			logger.V(1).Info("Cloud not close admin client")
```
Comment (saintstack) [inline controllers/update_status.go:335]: spelling

File: controllers/update_status.go [hunk 3]
```diff
@@ -306,49 +305,44 @@ func optionList(options ...string) []string {
 // tryConnectionOptions attempts to connect with all the combinations of
 // versions and connection strings for this cluster and returns the set that
 // allow connecting to the cluster.
```
Comment (saintstack) [inline controllers/update_status.go:307]: Does comment need updating now this method works differently?

**Review by saintstack — APPROVED** [review]
Some nits but otherwise LGTM.

**Review by brownleej — COMMENTED** [review]
How will this interact with the way the bindings cache the database connection?

**Review by saintstack — APPROVED** [review]

**Comment by johscheuer** [comment]
> How will this interact with the way the bindings cache the database connection?

We probably have the same limitations like before. I have to double check the c/go bindings what happens if we reuse a database struct to ensure the cluster file content will be read again. Otherwise we don't add more limitations than before, so I would go ahead and merge this PR. I still will validate/read the c/go bindings for my own education (and curiosity).

**Pull request merged**